akka-projection-core/src/main/scala/akka/projection/TelemetrySpi.scala

```diff
@@ -95,7 +102,9 @@ trait Telemetry {
 
   override def stopped(): Unit = {}
 
-  override def afterProcess(readyTimestampNanos: Long): Unit = {}
+  override def beforeProcess(): AnyRef = Unit
+
+  override def afterProcess(externalContext: AnyRef): Unit = {}
```

Comment:
I think we can only propagate `externalContext` to `afterProcess` (which is invoked only when successfully processing an envelope).

There are some locations where we invoke `error` (when even processing errors) in which we have no access to the context. I guess we could add `Optional[AnyRef]` and propagate the `externalContext` when available.

Reply:
That’s only for flow handler, right? I think we could skip using error completely for flow, because flow is power user area and not guaranteed that it will call afterProcess either. For flow it’s more interesting to look at failures, which will be equal to the number of errors.

Reply:
Hmm, I looked in more detail and I think propagating the `externalContext` in case of `error` is only possible in `SingleHandlerStrategy`.

 - in the case of `FlowHandlerStrategy` (see @patriknw's above)
 - in the case of `GroupHandlerStrategy` we would need to know which envelope in the group caused the error, or support an `error` overload that accepted a collection of `externalContext` and report them all (even if some envelopes succeeded and some were not even attempted)

I think we can revisit the need to propagate `externalContext` in `error` again in the future and, for now, stay with `def error(throwable)`.